pupil-spa/src/app/familiarisation-colour/familiarisation-colour.component.ts

```diff
@@ -5,38 +5,97 @@ import {
+import { AzureQueueService } from '../services/azure-queue/azure-queue.service';
+import { Pupil } from '../pupil';
+import { AuditService } from '../services/audit/audit.service';
 
 @Component({
   selector: 'app-familiarisation-colour',
   templateUrl: './familiarisation-colour.component.html',
   styleUrls: ['./familiarisation-colour.component.scss']
 })
 export class FamiliarisationColourComponent {
-  selectedContrast;
+  accessArrangements;
   contrastSettings;
+  featureUseHpa;
+  pupil: Pupil;
+  selectedContrast;
+  pupilPrefsAPIErrorDelay;
+  pupilPrefsAPIErrorMaxAttempts;
 
   constructor(
-    private router: Router,
+    private auditService: AuditService,
+    private azureQueueService: AzureQueueService,
     private storageService: StorageService,
-    private routeService: RouteService
+    private router: Router,
+    private routeService: RouteService,
+    private tokenService: TokenService,
+
   ) {
+    const { featureUseHpa, pupilPrefsAPIErrorDelay, pupilPrefsAPIErrorMaxAttempts } = APP_CONFIG;
+    this.accessArrangements = new AccessArrangements;
+    this.featureUseHpa = featureUseHpa;
     this.contrastSettings = AccessArrangementsConfig.contrastSettings;
+    this.pupilPrefsAPIErrorDelay = pupilPrefsAPIErrorDelay;
+    this.pupilPrefsAPIErrorMaxAttempts = pupilPrefsAPIErrorMaxAttempts;
+    const pupilData = storageService.getItem('pupil');
+    this.pupil = new Pupil;
+    this.pupil.checkCode = pupilData.checkCode;
+    const config = this.storageService.getItem('config');
+    const colourContrastSetting = this.contrastSettings.find(f => f.code === config.colourContrastCode);
+    this.selectedContrast = (colourContrastSetting && colourContrastSetting.val) || 'bow';
+    this.setColourContrast(this.selectedContrast);
+  }
+
+  setColourContrast(colourContrastValue) {
+    this.accessArrangements.contrast = colourContrastValue;
+    this.storageService.setItem(accessArrangementsDataKey, this.accessArrangements);
   }
 
   selectionChange(selectedContrast) {
     this.selectedContrast = selectedContrast;
   }
 
-  onClick() {
-    const accessArrangements = new AccessArrangements;
-    accessArrangements.contrast = this.selectedContrast || 'bow';
-    this.storageService.setItem(accessArrangementsDataKey, accessArrangements);
+  async onClick() {
+    this.setColourContrast(this.selectedContrast);
+    const colourContrastSetting = this.contrastSettings.find(f => f.val === this.accessArrangements.contrast);
+    const colourContrastCode = colourContrastSetting.code;
+    await this.submitAzureQueueMessage(colourContrastCode);
+    this.storageService.setItem(accessArrangementsDataKey, this.accessArrangements);
     if (this.routeService.getPreviousUrl() === '/access-settings') {
       this.router.navigate(['access-settings']);
     } else {
       this.router.navigate(['sign-in-success']);
     }
   }
+
+  async submitAzureQueueMessage(colourContrastCode) {
+    if (this.featureUseHpa === true) {
```

Comment:
Can this be `if (!this.featureUseHpa === true) return` to avoid the unnecessary nesting? Separating to another function as Jon suggested might also be good.

Reply:
@danwilliams extracted into a separate service and avoided nesting as suggested.